[clang][analyzer] Fix a crash in alpha.unix.BlockInCriticalSection

gamesh411 opened the pull request:
When analyzing C code with function pointers the checker crashes because
of how the implementation extracts IdentifierInfo. Without the fix, this
test crashes.

gamesh411 committed:
[clang][analyzer] Fix alpha.unix.BlockInCriticalSection

When analyzing C code with function pointers the checker crashes because
of how the implementation extracts IdentifierInfo. Without the fix, this
test crashes.

Add crashing test

steakhal reviewed (changes requested):
> In CTU there is not always an AnalysisDeclContext for a given call.

Why?

Could you demonstrate the fix in a test?

gamesh411 commented:
After reducing a crashing TU, I have found, that the issue came up without CTU analysis as well.
I have added a test case that demonstrates the crash without the fix.
I also updated the commit message to reflect the real cause.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp:
```diff
@@ -103,9 +104,10 @@ class RAIIMutexDescriptor {
-      Guard = &Call.getCalleeAnalysisDeclContext()->getASTContext().Idents.get(
-          GuardName);
-      IdentifierInfoInitialized = true;
+      if (AnalysisDeclContext *CalleCtx = Call.getCalleeAnalysisDeclContext()) {
+        Guard = &CalleCtx->getASTContext().Idents.get(GuardName);
+        IdentifierInfoInitialized = true;
+      }
```
This crash happened because we failed to get an ASTContext.
Maybe, on some other access-path we could acquire the ASTContext, like via `Call.getState()->getStateManager().getContext()`.

I feel like your proposal only fixes the symptom, and not the root cause.

steakhal reviewed (changes requested):
Now the change makes sense. I disagree with the fix though, see inline.

gamesh411 committed:
Get ASTContext through StateManager

steakhal commented on clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp:
```diff
@@ -14,6 +14,7 @@
 //
 //===----------------------------------------------------------------------===//
 
+#include "clang/Analysis/AnalysisDeclContext.h"
```
```suggestion
```
You can probably drop this line now.

steakhal reviewed (approved):
LGTM, thanks!

steakhal committed:
Remove unused include
